File: C++/3026-Maximum Good Subarray Sum.cpp
using LL = long long;
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        unordered_map<int, LL> hmap;  // number:minimum preSum
        LL preSum = 0;
        LL res = LLONG_MIN;
        for (int num:nums) {
            preSum += num;
            if (hmap.find(num-k) != hmap.end()) {
                res = max(res, preSum - hmap[num-k]);
            }
            if (hmap.find(num+k) != hmap.end()) {
                res = max(res, preSum - hmap[num+k]);
            }

            if (hmap.find(num) != hmap.end()) {
                hmap[num] = min(hmap[num], preSum-num);
            } else {
                hmap[num] = preSum-num;
            }
            
        }
        return res>LLONG_MIN?res:0;
    }
};

File: 3026-Maximum Good Subarray Sum.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        hmap = defaultdict(lambda:float('inf'))  # num:preSum
        preSum = 0
        res = float('-inf')
        for i in range(len(nums)):
            if nums[i] - k in hmap:
                res = max(res, preSum + nums[i] - hmap[nums[i]-k])
            if nums[i] + k in hmap:
                res = max(res, preSum + nums[i] - hmap[nums[i]+k])
            hmap[nums[i]] = min(hmap[nums[i]], preSum)
            preSum += nums[i]

        return res if res > float('-inf') else 0